merkle/merkle_tree.go

```diff
@@ -0,0 +1,196 @@
+/*
+Package merkle provides functions and types to deal with the creation and storage of
+Merkle trees, using the secure SHAKE256 KDF implemented in the signing/key package:
+thus not being affected by the the infamous M-Bug.
+
+The functions exported by the package are:
+
+- CreateMerkleTree(seed trinary.Hash, securityLvl int, depth int) creates a MerkleTree
+  structure of the specified depth, using a SHAKE256 key of the the length specified by
+  the supplied securitylevel, deriving subseeds from the provided seed.
+- StoreMerkleTreeFile(filePath string, merkleTree *MerkleTree) stores the MerkleTree
+  structure in a file; the format used is compatible with Hornet.
+- LoadMerkleTreeFile(filePath string) loads a Hornet-compatible Merkle tree file as a
+  MerkleTree structure.
+*/
+package merkle
+
+import (
+	"math"
+	"runtime"
+	"sync"
+
+	"github.com/iotaledger/iota.go/consts"
+	"github.com/iotaledger/iota.go/kerl"
+	"github.com/iotaledger/iota.go/signing"
+	"github.com/iotaledger/iota.go/signing/key"
+	"github.com/iotaledger/iota.go/trinary"
+)
+
+// MerkleTree contains the merkle tree used for the coordinator signatures.
+type MerkleTree struct {
+	// depth of the merkle tree
+	Depth int
+	// root address of the merkle tree
+	Root trinary.Hash
+	// merkle tree layers indexed by their level
+	Layers map[int]*MerkleTreeLayer
+}
+
+// MerkleTreeLayer contains the nodes of a layer of a merkle tree.
+type MerkleTreeLayer struct {
+	// level of the layer in the tree
+	Level int
+	// nodes of the layer
+	Hashes []trinary.Hash
+}
+
+type MilestoneIndex uint32
+
+// calculateAllAddresses calculates all addresses that are used for the merkle tree of the coordinator.
+func calculateAllAddresses(seed trinary.Hash, securityLvl int, count int) []trinary.Hash {
+	resultLock := &sync.Mutex{}
+	result := make([]trinary.Hash, count)
+
+	wg := sync.WaitGroup{}
+	wg.Add(runtime.NumCPU())
```

Comment:
I've actually realized that `iota.go` already uses `runtime.NumCPU` in `pow_go.go`. Should we keep it?
Regarding allowing the user to choose the level of parallelism of the merkle creation tool I think it is overkill: not for the tool itself, but for the library to account for such a user decision; it would be true for PoW too at the moment, a way more frequent operation.

Reply:
I was actually proposing to make it configurable, because the PoW has an optional parameter like this.
I really don't think that `runtime.NumCPU()` is a good idea, because it always only returns the number of logical CPUs independently of what could actually be used. `GOMAXPROCS` is more flexible here, and I guess the user can also override it with an environmental variable. Which might indeed render it unnecessary to have this a parameter :thinking: 
@luca-moser what do you think?

Reply:
I think as with the PoW funcs, we should add an optional `paralleism ...int` parameter in case one wants to use a specific amount of concurrent goroutines for the Merkle tree generation. Per default it should use `runtime.GOMAXPROCS(0)` and not `runtime.NumCPU()` since `GOMAXPROCS` could have been changed and be different from `runtime.NumCPU()` (its default value however is the same as `runtime.NumCPU()`).